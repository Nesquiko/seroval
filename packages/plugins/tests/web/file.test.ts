import {
  crossSerializeAsync,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serializeAsync,
  toCrossJSONAsync,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import FilePlugin from '../../web/file';

const EXAMPLE = new File(['Hello World'], 'hello.txt', {
  type: 'text/plain',
  lastModified: 1681027542680,
});

describe('File', () => {
  describe('serializeAsync', () => {
    it('supports File', async () => {
      const result = await serializeAsync(EXAMPLE, {
        plugins: [FilePlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back).toBeInstanceOf(File);
      expect(await back.text()).toBe(await EXAMPLE.text());
      expect(back.type).toBe(EXAMPLE.type);
    });
  });
  describe('toJSONAsync', () => {
    it('supports File', async () => {
      const result = await toJSONAsync(EXAMPLE, {
        plugins: [FilePlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result, {
        plugins: [FilePlugin],
      });
      expect(back).toBeInstanceOf(File);
      expect(await back.text()).toBe(await EXAMPLE.text());
      expect(back.type).toBe(EXAMPLE.type);
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports File', async () => {
      const result = await crossSerializeAsync(EXAMPLE, {
        plugins: [FilePlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports File', async () => {
        const result = await crossSerializeAsync(EXAMPLE, {
          plugins: [FilePlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports File', async () => {
      const result = await toCrossJSONAsync(EXAMPLE, {
        plugins: [FilePlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        plugins: [FilePlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(File);
      expect(await back.text()).toBe(await EXAMPLE.text());
      expect(back.type).toBe(EXAMPLE.type);
    });
  });
});
