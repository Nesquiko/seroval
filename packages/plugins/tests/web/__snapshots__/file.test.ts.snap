// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`File > crossSerializeAsync > scoped > supports File 1`] = `"($R=>$R[0]=new File([$R[2]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],"hello.txt",$R[1]={type:"text/plain",lastModified:1681027542680}))($R["example"])"`;

exports[`File > crossSerializeAsync > supports File 1`] = `"$R[0]=new File([$R[2]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],"hello.txt",$R[1]={type:"text/plain",lastModified:1681027542680})"`;

exports[`File > serializeAsync > supports File 1`] = `"new File([new Uint8Array([72,101,108,108,111,32,87,111,114,108,100]).buffer],"hello.txt",{type:"text/plain",lastModified:1681027542680})"`;

exports[`File > toCrossJSONAsync > supports File 1`] = `"{"t":40,"i":0,"s":{"name":{"t":1,"s":"hello.txt"},"options":{"t":10,"i":1,"p":{"k":["type","lastModified"],"v":[{"t":1,"s":"text/plain"},{"t":0,"s":1681027542680}],"s":2},"o":0},"buffer":{"t":21,"i":2,"s":[72,101,108,108,111,32,87,111,114,108,100]}},"c":"seroval-plugins/web/url"}"`;

exports[`File > toJSONAsync > supports File 1`] = `"{"t":{"t":40,"i":0,"s":{"name":{"t":1,"s":"hello.txt"},"options":{"t":10,"i":1,"p":{"k":["type","lastModified"],"v":[{"t":1,"s":"text/plain"},{"t":0,"s":1681027542680}],"s":2},"o":0},"buffer":{"t":21,"i":2,"s":[72,101,108,108,111,32,87,111,114,108,100]}},"c":"seroval-plugins/web/url"},"f":16383,"m":[]}"`;
